Azure/sonic-utilities
# [config] Add 'bgp shutdown\startup neighbor <ip>' subcommands for FRR

[frr_config_bgp.py]: New File with BGP group of python click.

Signed-off-by: Praveen Chaudhary<pchaudhary@linkedin.com>

Changes:
1.) import bgp group from frr_config_bgp.py for frr routing stack.
2.) new file frr_config_bgp.py to have bgp group related config commands.



**- What I did**
[main.py]: Add 'Config bgp shutdown\startup neighbor <ip>' for FRR.

**- How I did it**
Changes:
1.) import bgp group from frr_config_bgp.py for frr routing stack.
2.) new file frr_config_bgp.py to have bgp group related config commands.


**- How to verify it**
Testing on DUT:

```
— Test successful path:
A.) Make sure sudo is specified before config bgp
admin@falco-test-dut01:~$ config bgp shutdown neighbour
Root privileges are required for this operation
B.) Command does not ask for sudo PWD.
admin@falco-test-dut01:~$ sudo config bgp shutdown neighbor 10.0.0.3
or
admin@falco-test-dut01:~$ sudo config bgp shutdown neighbor 10.0.0.3 -v
Running command: sudo vtysh -c 'show bgp neighbor 10.0.0.3' | grep "local AS"
local ASN for neighbor 10.0.0.3 is 65100
Running command: sudo vtysh -c 'configure terminal' -c 'router bgp 65100' -c 'neighbor 10.0.0.3 shutdown'
Verify:
Command: sudo vtysh -c "show bgp summary"
IPv4 Unicast Summary:
BGP router identifier 10.1.0.32, local AS number 65100 vrf-id 0
BGP table version 12196
RIB entries 13087, using 1738 KiB of memory
Peers 64, using 1331 KiB of memory
Peer groups 2, using 144 bytes of memory
Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
10.0.0.1        4      65200    3633    3676        0    0    0 00:21:33         6402
10.0.0.3        4      65200    3573    3616        0    0    0 00:03:02 Idle (Admin)<<<<<

Startup:
admin@falco-test-dut01:~$ sudo config bgp startup neighbor 10.0.0.3
Or

admin@falco-test-dut01:~$ sudo config bgp startup neighbor 10.0.0.3 -v
Running command: sudo vtysh -c 'show bgp neighbor 10.0.0.3' | grep "local AS"
local ASN for neighbor 10.0.0.3 is 65100
Running command: sudo vtysh -c 'configure terminal' -c 'router bgp 65100' -c 'no neighbor 10.0.0.3 shutdown'

Verify: 
Command: sudo vtysh -c "show bgp summary"
IPv4 Unicast Summary:
BGP router identifier 10.1.0.32, local AS number 65100 vrf-id 0
BGP table version 12197
RIB entries 13089, using 1738 KiB of memory
Peers 64, using 1331 KiB of memory
Peer groups 2, using 144 bytes of memory
Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
10.0.0.1        4      65200    3683    3727        0    0    0 00:24:04         6402
10.0.0.3        4      65200    6787    6866        0    0    0 00:00:31         6402
admin@lnos-x1-a-csw04:~$ sudo config bgp startup all
NO-OP Command for Sonic FRR
admin@lnos-x1-a-csw04:~$ sudo config bgp startup neighbor 10.0.0.3
Error: neighbor '10.0.0.3' is not found
Aborted!

Latest Test Results:
Ipv6:
admin@falco-test-dut01:~$ sudo config bgp shutdown neighbor fc00::72
admin@falco-test-dut01:~$
Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
fc00::72        4      64600   16198   16154        0    0    0 00:00:10 Idle (Admin)
fc00::76        4      64600   16201   18992        0    0    0 10:49:58         6002
fc00::7a        4      64600   16202   19010        0    0    0 10:49:58         6002
fc00::7e        4      64600   16202   18843        0    0    0 10:49:56         6002
admin@falco-test-dut01:~$ sudo config bgp startup neighbor fc00::72
admin@falco-test-dut01:~$ show bgp summary
Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
fc00::72        4      64600   19409   19170        0    0    0 00:00:22         6002
fc00::76        4      64600   16239   19031        0    0    0 10:51:51         6002
fc00::7a        4      64600   16239   19048        0    0    0 10:51:51         6002
fc00::7e        4      64600   16240   18882        0    0    0 10:51:49         6002
admin@falco-test-dut01:/etc/sonic$ sudo config bgp shutdown neighbor fc00::74
Error: BGP information not found for Neighbor fc00::74
Aborted!
Ipv4:
admin@falco-test-dut01:/etc/sonic$ sudo config bgp shutdown neighbor 10.0.0.3
Error: BGP information not found for Neighbor 10.0.0.3
Aborted!
admin@falco-test-dut01:/etc/sonic$ sudo config bgp shutdown neighbor 10.0.0.57
admin@falco-test-dut01:/etc/sonic$
admin@falco-test-dut01:/etc/sonic$ show bgp summary 
Command: sudo vtysh -c "show bgp summary"
IPv4 Unicast Summary:
BGP router identifier 10.1.0.32, local AS number 65100 vrf-id 0
BGP table version 11429
RIB entries 12013, using 1595 KiB of memory
Peers 8, using 166 KiB of memory
Peer groups 3, using 216 bytes of memory
Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
10.0.0.57       4      64600    8420    8225        0    0    0 00:00:19 Idle (Admin)
10.0.0.59       4      64600    8427    8610        0    0    0 04:21:12         6002
10.0.0.61       4      64600    8428    8610        0    0    0 04:21:12         6002
10.0.0.63       4      64600    8427    8231        0    0    0 04:21:11         6002

admin@falco-test-dut01:/etc/sonic$ sudo config bgp startup neighbor 10.0.0.3
Error: BGP information not found for Neighbor 10.0.0.3
Aborted!
admin@falco-test-dut01:/etc/sonic$ sudo config bgp startup neighbor 10.0.0.57
admin@falco-test-dut01:/etc/sonic$ 
admin@falco-test-dut01:/etc/sonic$ show bgp summary 
Command: sudo vtysh -c "show bgp summary"
IPv4 Unicast Summary:
BGP router identifier 10.1.0.32, local AS number 65100 vrf-id 0
BGP table version 11430
RIB entries 12015, using 1596 KiB of memory
Peers 8, using 166 KiB of memory
Peer groups 3, using 216 bytes of memory
Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
10.0.0.57       4      64600   11626   11234        0    0    0 00:00:08         6002
10.0.0.59       4      64600    8442    8626        0    0    0 04:21:58         6002
10.0.0.61       4      64600    8443    8626        0    0    0 04:21:58         6002
10.0.0.63       4      64600    8442    8247        0    0    0 04:21:57         6002
```
**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**
